electrum/wallet.py

```diff
@@ -1061,16 +1053,25 @@ def _init_requests_rhash_index(self):
                 self._requests_addr_to_rhash[addr] = req.rhash
 
     def _prepare_onchain_invoice_paid_detection(self):
-        # scriptpubkey -> list(invoice_keys)
-        self._invoices_from_scriptpubkey_map = defaultdict(set)  # type: Dict[bytes, Set[str]]
+        self._invoices_from_txid_map = defaultdict(set)  # type: Dict[str, Set[str]]
+        self._invoices_to_txid_map = defaultdict(set)  # type: Dict[str, Tuple[bool, Optional[int], Sequence[str]]]
+        self._detect_paid_invoices()
+
+    def _detect_paid_invoices(self):
         for invoice_key, invoice in self._invoices.items():
```

Comment:
This iterates over all invoices, and is called in `add_transaction`. This is a performance regression that we should avoid IMHO. Instead, I propose: https://github.com/spesmilo/electrum/pull/7942